src/python/pants/build_graph/address.py

```diff
@@ -233,10 +227,31 @@ def __init__(
         """
         self.spec_path = spec_path
         self._relative_file_path = relative_file_path
+
         # If the target_name is the same as the default name would be, we normalize to None.
-        self._target_name = (
-            target_name if target_name and target_name != os.path.basename(self.spec_path) else None
-        )
+        self._target_name: Optional[str]
+        if target_name and target_name != os.path.basename(self.spec_path):
+            banned_chars = BANNED_CHARS_IN_TARGET_NAME & set(target_name)
+            deprecated_banned_chars = banned_chars & set(r"/\:")
+            if deprecated_banned_chars:
+                warn_or_error(
+                    removal_version="2.2.0.dev1",
+                    deprecated_entity_description=r"Using any of the `\`, `/`, or `:` characters in a target name.",
+                    hint=(
+                        "The target name {target_name}, defined in directory {self.spec_path} "
+                        "contains deprecated characters (`{banned_chars}`). Please replace these "
+                        "characters with another separator character like `_` or `-`."
+                    ),
+                )
+            elif banned_chars:
+                raise InvalidTargetName(
+                    f"Banned chars found in target name. {banned_chars} not allowed in target "
+                    f"name: {target_name}"
```

Comment:
Needs to mention the `self.spec_path` I think. I like the wording in the deprecation warning.

Reply:
Sure.